### PR TITLE
EZP-31440: Simplified non content/location related Query Types rendering

### PR DESCRIPTION
**Pull request by adamwojs**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31440](https://jira.ez.no/browse/EZP-31440)
| **Bug/Improvement**| no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

Introduced `ez_render_<type>_query`  and `ez_render_<type>_query_<renderer>` Twig functions to simplify  (no need to create custom controller) non content/location related Query Types rendering. Example use cases:

* Rendering "Most recent posts" section in the sidebar on blog post view containing 
* Rendering "Top rated articles" section in website footer
* Rendering category menu
 
### Minimal

```twig
{{ ez_render_content_query({
    'query': {
        'query_type': 'LatestBlogPost',
    },
    'template': 'latest_blog_post.html.twig',
}) }}
```

### All options

```twig
{{ ez_render_location_query_esi({
    'query': {
        'query_type': 'LatestArticle',
        'parameters': {
            'category': 'games'
        },
        'assign_results_to': 'latest_articles'
    },
    'pagination': {
        'enebled': true,
        'limit': 10,
        'page_param': 'page'
    }
    'template': 'latest_articles.html.twig',
}) }}
```

**TODO**:
- [X] Implement feature / fix a bug.
- [x] Implement tests.
- [X] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by webhdx** [comment]
Please ping me when you are done with implementing tests.

**Comment by adamwojs** [comment]
> Please ping me when you are done with implementing tests.

@webhdx Sure. Is it OK from conceptual POV for you?  (that's why I asked for feedback here)

**Comment by webhdx** [comment]
Yes, I like the idea 👍

**Comment by adamwojs** [comment]
> I'm sorry to say that in general I don't like this approach. I had experience with this issue when doing my own project. What I expected as a developer was to be able to re-use things that are already given by Symfony: 
>
> [...] 
> 
> The reasoning behind this expectation is actually quite simple - ez_query:* are Controllers, so we should be able to use them as Controllers.

And where is the issue? You are still able to do so if you like. 

```
{{ ez_render_location_query_esi({
    'query': {
        'query_type': 'LatestBlogPost',
    },
    'template': 'latest_blog_post.html.twig',
}) }}
```

is a shorthand for:

```
{{ render_esi(
    controller(
        'ez_query_renderer:locationQueryAction',
        {
            'query': {
                'query_type': 'LatestBlogPost',
            },
           'template': 'latest_blog_post.html.twig'
        }
) })
```

> What is missing for the above to work is a View Builder which would provide ContentView parameter to that controller when used in this way. I tried to do it quickly but it seems it's more tricky. Seems to me it's possible however (correct me if I'm missing something).

Please read PR description carefully. The point of the whole feature is to simplify rendering of NON content/location related Query Types. How would you like to provide  `ContentView` for something which is by defintion not related to any content ?

**Comment by adamwojs** [comment]
@webhdx Unit tests added in https://github.com/ezsystems/ezpublish-kernel/pull/2970/commits/0d00028ef4588299ffa30c9b2fb665e9fc4e39fa

File: eZ/Bundle/EzPublishCoreBundle/Resources/config/templating.yml [hunk 1]
```diff
@@ -155,6 +155,12 @@ services:
         tags:
             - { name: twig.extension }
 
+    eZ\Publish\Core\MVC\Symfony\Templating\Twig\Extension\QueryRenderingExtension:
+        arguments:
+            - "@fragment.handler"
```
Comment (alongosz) [inline eZ/Bundle/EzPublishCoreBundle/Resources/config/templating.yml]: CS
```suggestion
            - '@fragment.handler'
```

File: eZ/Publish/Core/Pagination/Pagerfanta/AdapterFactory/SearchHitAdapterFactory.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    public function __construct(SearchService $searchService)
+    {
+        $this->searchService = $searchService;
+    }
+
+    public function createAdapter(Query $query, array $languageFilter = []): AdapterInterface
+    {
+        if ($query instanceof LocationQuery) {
+            return new LocationSearchHitAdapter($query, $this->searchService, $languageFilter);
+        } else {
```
Comment (ViniTou) [inline eZ/Publish/Core/Pagination/Pagerfanta/AdapterFactory/SearchHitAdapterFactory.php]: not rly needed

Comment (adamwojs) [reply eZ/Publish/Core/Pagination/Pagerfanta/AdapterFactory/SearchHitAdapterFactory.php]: Fixed in https://github.com/ezsystems/ezpublish-kernel/pull/2970/commits/c0d91a5c05a1e8c6c92097b16e9555ee55d76189

**Comment by adamwojs** [comment]
Note for reviewers: I had to remove `ez_render_invalid_query.test` becouse it was OS dependent. See   https://ci.appveyor.com/project/ezrobot/ezpublish-kernel/builds/31446183 

_I didn't notice this before because for some reason AppVeyor is not triggered after `git push`_

**Comment by gggeek** [comment]
Hello, I just gave the code a quick scan, but I did not find any trace of TTL/content-based-expiration controls for these 'query views'. Am I correct?
To me, that is the most important (and complex) part of query views...

**Comment by andrerom** [comment]
Assuming these are rendered inline they will add tags to http cache, however you are right that there is a lack of TTL or other mechanism to handle cases where the results of the query changes as opposed to just invalidation on changes to the currently rendered content.

If we wanted to deal with this using invalidation (and not ttl expiry), then we probably need to do reverse lookup for all registered queries on post publish of content to validate if query results are affected or not, then some special http cache tags for this to be able to invalidate on.

**Comment by gggeek** [comment]
@andrerom agree :-)
- add automatically as tags the ids of the contents (locations?) matched in this query => good!
- bring back the 'smart viewcache' of olde (as in f.e.: 'query all-latest-contents-in-location-X should expire when new content is added to that location) => better, but it might both be complex in code and limited in expressive power

suggestions:

- add a way to express a simple ttl as param for the view-query => not ideal, but can cover a lot of complex scenarios in an 'aceptable' way

- future-work: add a more-flexible reimplementation of smart-view-cache, where users can add a cache-tag-id to any controller they want, and somehow "easily" add 'invalidate this cache tag' rules for events such as publish-content-of-type-x, add-or-remove-contents-in-location-y, etc (atm it could be probably be done use kaliop workflowengine bundle, but config would be overkill for all but hardcode devs)

**Pull request merged**